pkg/compactor/blocks_cleaner.go

```diff
@@ -402,17 +411,28 @@ func (c *BlocksCleaner) deleteBlocksMarkedForDeletion(ctx context.Context, idx *
 
 // cleanUserPartialBlocks delete partial blocks which are safe to be deleted. The provided partials map
 // and index are updated accordingly.
-func (c *BlocksCleaner) cleanUserPartialBlocks(ctx context.Context, partials map[ulid.ULID]error, idx *bucketindex.Index, userBucket objstore.InstrumentedBucket, userLogger log.Logger) {
+func (c *BlocksCleaner) cleanUserPartialBlocks(ctx context.Context, partials []bucketindex.PartialBlock, idx *bucketindex.Index, delay time.Duration, userBucket objstore.InstrumentedBucket, userLogger log.Logger) {
 	// Collect all blocks with missing meta.json into buffered channel.
 	blocks := make([]ulid.ULID, 0, len(partials))
 
-	for blockID, blockErr := range partials {
+	for _, p := range partials {
 		// We can safely delete only blocks which are partial because the meta.json is missing.
-		if !errors.Is(blockErr, bucketindex.ErrBlockMetaNotFound) {
+		if !errors.Is(p.Reason, bucketindex.ErrBlockMetaNotFound) {
 			continue
 		}
-
-		blocks = append(blocks, blockID)
+		// Check if partial block is older than delay period, and mark for deletion if necessary
+		if delay > 0 {
+			maxTime := time.Unix(p.MaxTime/1000, 0)
```

Comment:
We don't want to base this on block's MaxTime. Instead we want to list all files in the block, and find Maximum "Last Modified" time from all the files. We can find "Last Modified" time using `bucket.Attributes()` call.

> We don't want to base this on block's MaxTime

This wouldn't work very well -- user may be uploading (backfilling) old blocks, with MaxTime being way in the past.

Reply:
I added a function to calculate the last modified time for all files given a bucket and blockid and replaced this code in the latest commit